src/protocol.rs

```diff
@@ -121,145 +131,70 @@ impl Participant {
         &mut self,
         message: &Message,
         rng: &mut R,
-    ) -> Result<(Identifier, Output, Vec<Message>)> {
+    ) -> Result<(Identifier, Option<P::Output>, Vec<Message>)> {
         info!("Processing single message.");
 
         if message.to() != self.id {
             return Err(InternalError::WrongMessageRecipient);
         }
-        match message.message_type() {
-            MessageType::Auxinfo(_) => {
+        if message.id() != self.sid {
+            error!(
+                "Message for session {} was routed to the wrong participant (sid: {})!",
+                message.id(),
+                self.sid
+            );
+            return Err(InternalError::ProtocolError);
+        }
+        match (message.message_type(), P::protocol_type()) {
```

Comment:
I think this could be simplified: the match stays the same, but the body of the "acceptable" arms goes to nothing, since they're all the same. If this doesn't throw an error, then we call `process message` and handle the outcome identically for all `P`.